cephfs/admin/subvolume_test.go

```diff
@@ -156,6 +156,66 @@ func TestRemoveSubVolume(t *testing.T) {
 			return fsa.RemoveSubVolumeWithFlags(v, g, n, SubVolRmFlags{RetainSnapshots: true})
 		})
 	})
+	t.Run("retainedSnapshotsTest", func(t *testing.T) {
+		subname := "retsnap1"
+		snapname := "s1"
+		err = fsa.CreateSubVolume(volume, NoGroup, subname, nil)
+		assert.NoError(t, err)
+		vinfo, err := fsa.SubVolumeInfo(volume, NoGroup, subname)
+		assert.NoError(t, err)
+
+		canRetain := false
+		for _, f := range vinfo.Features {
+			if f == SnapshotRetentionFeature {
+				canRetain = true
+			}
+		}
+		if !canRetain {
+			err = fsa.RemoveSubVolumeWithFlags(
+				volume, NoGroup, subname, SubVolRmFlags{Force: true})
+			assert.NoError(t, err)
+			t.Skipf("this rest of this test requires snapshot retention on the server side")
+		}
+
+		lsv, err = fsa.ListSubVolumes(volume, NoGroup)
+		assert.NoError(t, err)
+		afterCount := len(lsv)
+		assert.Equal(t, beforeCount, afterCount-1)
+		err = fsa.CreateSubVolumeSnapshot(volume, NoGroup, subname, snapname)
+
+		err = fsa.RemoveSubVolumeWithFlags(
+			volume, NoGroup, subname, SubVolRmFlags{Force: true})
+		assert.Error(t, err)
+
+		err = fsa.RemoveSubVolumeWithFlags(
+			volume, NoGroup, subname, SubVolRmFlags{RetainSnapshots: true})
+		assert.NoError(t, err)
+
+		delay()
+		_, err = fsa.SubVolumeInfo(volume, NoGroup, subname)
+		assert.NoError(t, err)
+
+		err = fsa.RemoveSubVolumeSnapshot(volume, NoGroup, subname, snapname)
+		assert.NoError(t, err)
+		err = fsa.RemoveSubVolumeWithFlags(
```

Comment:
question. do we need to remove the subvolume again or ceph will garbage collect it once all the snapshots on that subvolume are deleted?

Reply:
Good question, but I don't have a proper answer. I wasn't patient enough to see if simply waiting a while would make the volume go away.... but even after calling for this subsequent delete it still can take a while. *shrugs* I'd check with a real cephfs expert if this question matters to csi.

Reply:
Well, that's the bug. The underscore matters to the JSON but not to the CLI. The CLI converts the dash to an underscore when constructing the JSON.